lib/bolt_server/base_config.rb

```diff
@@ -0,0 +1,140 @@
+                                'ECDHE-ECDSA-AES128-SHA256',
+                                'ECDHE-RSA-AES128-SHA256'],
+        'loglevel' => 'notice',
+        'file-server-conn-timeout' => 120 }
+    end
+
+    def self.config_keys
+      raise NotImplementedError, "self.config_keys() must be implemented by the transport class"
+    end
+
+    def self.env_keys
+      raise NotImplementedError, "self.env_keys() must be implemented by the transport class"
+    end
+
+    def self.defaults
+      self.class.base_defaults
+    end
+
+    def initialize(config = nil)
+      @data = self.class.defaults
+      @data = @data.merge(config.select { |key, _| self.class.config_keys.include?(key) }) if config
+      @config_path = nil
+    end
+
+    def load_file_config(path)
+      @config_path = path
+      begin
+        # This lets us get the actual config values without needing to
+        # know the service name
+        parsed_hocon = Hocon.load(path).values[0]
+      rescue Hocon::ConfigError => e
+        raise "Hocon data in '#{path}' failed to load.\n Error: '#{e.message}'"
+      rescue Errno::EACCES
+        raise "Your user doesn't have permission to read #{path}"
+      end
+
+      raise "Could not find service config at #{path}" if parsed_hocon.nil?
+
+      parsed_hocon = parsed_hocon.select { |key, _| self.class.config_keys.include?(key) }
+
+      @data = @data.merge(parsed_hocon)
+    end
+
+    def load_env_config
+      self.class.env_keys.each do |key|
+        transformed_key = "BOLT_#{key.tr('-', '_').upcase}"
+        next unless ENV.key?(transformed_key)
+        @data[key] = if ['concurrency', 'file-server-conn-timeout'].include?(key)
+                       ENV[transformed_key].to_i
+                     else
+                       ENV[transformed_key]
+                     end
+      end
+    end
+
+    def natural?(num)
+      num.is_a?(Integer) && num.positive?
+    end
+
+    def ssl_keys
+      ['ssl-cert', 'ssl-key', 'ssl-ca-cert']
+    end
+
+    def required_keys
+      ssl_keys + ['file-server-uri']
+    end
+
+    def validate
+      required_keys.each do |k|
+        next unless @data[k].nil?
+        raise Bolt::ValidationError, "You must configure #{k} in #{@config_path}"
+      end
+
+      unless natural?(port)
+        raise Bolt::ValidationError, "Configured 'port' must be a valid integer greater than 0"
+      end
+      ssl_keys.each do |sk|
+        unless File.file?(@data[sk]) && File.readable?(@data[sk])
+          raise Bolt::ValidationError, "Configured #{sk} must be a valid filepath"
+        end
+      end
+
+      unless ssl_cipher_suites.is_a?(Array)
+        raise Bolt::ValidationError, "Configured 'ssl-cipher-suites' must be an array of cipher suite names"
+      end
+
+      unless whitelist.nil? || whitelist.is_a?(Array)
+        raise Bolt::ValidationError, "Configured 'whitelist' must be an array of names"
+      end
+
+      if self.class.config_keys.include?('concurrency')
```

Comment:
Would it be better to break this into a "natural numbers" list of configurations? I don't think that's any shorter and like having the example of service-specific validation, but having 2 in a row does seem redundant.